### PR TITLE
chore(cli): enable composite project

### PR DESCRIPTION
**Pull request by louis-bompart**
## Proposed changes

TypeScript server language in VS Code got mad cause of the rootDir + paths.
So this fix that by creating YATSCF (Yet-Another-TypeScript-Config-File)

Now we have:
- `tsconfig.dev.json` for running cli/plugin in dev.
- `tsconfig.build.json` to build the cli/plugin
- `tsconfig.json` for the IDE Language Serve

### CODE REVIEW

File: packages/cli/core/src/commands/ui/create/atomic.spec.ts [hunk 1]
```diff
@@ -115,6 +115,10 @@ describe('ui:create:atomic', () => {
     .do(() => {
       preconditionStatus.apiKey = false;
     })
+    .env({
+      TS_NODE_PROJECT:
+        '/Users/lbompart/Repos/cli/packages/cli/core/tsconfig.dev.json',
+    })
     .command(['ui:create:atomic', 'myapp'])
```
Comment (louis-bompart) [inline packages/cli/core/src/commands/ui/create/atomic.spec.ts:122]:
```suggestion
    .command(['ui:create:atomic', 'myapp'])
```

File: packages/cli/source/tsconfig.dev.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (y-lakhdar) [inline packages/cli/source/tsconfig.dev.json:1]: do we need to have one tsconfig.dev and tsconfig.build in the core/ folder?

**Comment by louis-bompart** [comment]
Update: the project is now composite.
The reason being is that if I just put `path`, typescript creeps around and bundles commons with core. (same for cli-plugin-source)

No impact tho on the build result with composite set properly

File: packages/cli/core/jest.config.js [hunk 3]
```diff
@@ -2,9 +2,15 @@ const {defaults} = require('jest-config');
 const {pathsToModuleNameMapper} = require('ts-jest');
 const {compilerOptions} = require('./tsconfig.json');
 
+process.env.TS_NODE_PROJECT = require.resolve('./tsconfig.dev.json');
```
Comment (louis-bompart) [inline packages/cli/core/jest.config.js:5]: tsconfig.dev is used here. and yep its required

File: packages/cli/source/tsconfig.dev.json (hunk 2 above)
Comment (louis-bompart) [reply packages/cli/source/tsconfig.dev.json:1]: https://github.com/coveo/cli/pull/913/files#r964405952

Yep, this tsconfig is used by oclif in dev mode, aka `bin/dev`.

File: .gitignore [hunk 4]
```diff
@@ -12,7 +12,7 @@ npm-debug.log
 reports/*.xml
 veracode
 artifacts
-
+*.tsbuildinfo
```
Comment (louis-bompart) [inline .gitignore:15]: this is generated by composite project and can be used for incremental rebuild (but we ain't there yet, hold ye 🐴 )

File: packages/cli-e2e/jest.config.ts [hunk 5]
```diff
@@ -5,7 +5,8 @@
 
 export default {
   testEnvironment: 'node',
-  globalSetup: process.env.CI ? './setup/ci.ts' : './setup/local.ts',
+  // globalSetup: process.env.CI ? './setup/ci.ts' : './setup/local.ts',
+  globalSetup: './setup/local.ts',
```
Comment (louis-bompart) [inline packages/cli-e2e/jest.config.ts:9]:
```suggestion
  globalSetup: process.env.CI ? './setup/ci.ts' : './setup/local.ts',
```

File: packages/cli-e2e/package.json [hunk 6]
```diff
@@ -23,7 +23,7 @@
     "test:headless": "node ./entrypoints/entry.js",
     "test:debug": "node ./entrypoints/entry.js --debug",
     "test:bash": "node ./entrypoints/entry.js --bash",
-    "jest": "jest --verbose --runInBand --detectOpenHandles --forceExit",
+    "jest": "npm run jest:ci",
```
Comment (louis-bompart) [inline packages/cli-e2e/package.json:26]: DRY

File: packages/cli-e2e/setup/ci.ts [hunk 7]
```diff
@@ -20,8 +20,9 @@ export default async function () {
   if (!(process.env.E2E_USE_NPM_REGISTRY === 'true')) {
     await startVerdaccio();
   }
+  process.env.npm_config_registry = 'http://localhost:4873';
   await installCli();
   process.stdout.write(`CLI PATH : ${process.env.CLI_EXEC_PATH}`);
   createUiProjectDirectory();
-  restoreCliConfig();
+  // restoreCliConfig();
```
Comment (louis-bompart) [inline packages/cli-e2e/setup/ci.ts:27]:
```suggestion
  restoreCliConfig();
```

File: packages/cli-e2e/setup/local.ts [hunk 8]
```diff
@@ -30,4 +32,5 @@ export default async function () {
   await npmLogin();
   scaffoldDummyPackages();
   await publishPackages();
+  await installCli();
```
Comment (louis-bompart) [inline packages/cli-e2e/setup/local.ts:35]:
```suggestion
  await publishPackages();
```

File: packages/cli-e2e/setup/local.ts [hunk 9]
```diff
@@ -22,6 +23,7 @@ export default async function () {
   useCIConfigIfEnvIncomplete();
   mkdirSync(SCREENSHOTS_PATH, {recursive: true});
   setProcessEnv();
+  process.env.npm_config_registry = 'http://localhost:4873';
```
Comment (louis-bompart) [inline packages/cli-e2e/setup/local.ts:26]:
```suggestion
  setProcessEnv();
```

File: packages/cli-e2e/setup/local.ts [hunk 10]
```diff
@@ -14,6 +14,7 @@ import {
   publishPackages,
   shimNpm,
   scaffoldDummyPackages,
+  installCli,
```
Comment (louis-bompart) [inline packages/cli-e2e/setup/local.ts:17]:
```suggestion
  scaffoldDummyPackages,
```

File: packages/cli/commons-dev/package.json [hunk 11]
```diff
@@ -16,7 +16,10 @@
     "./.prettierrc": "./.prettierrc.js",
     "./.prettierrc.js": "./.prettierrc.js",
     "./tsconfig.json": "./tsconfig.json",
-    "./*": "./lib/*"
+    "./*": {
+      "types": "./lib/*.d.ts",
+      "default": "./lib/*.js"
+    }
```
Comment (louis-bompart) [inline packages/cli/commons-dev/package.json:22]: https://nodejs.org/api/packages.html#community-conditions-definitions
https://nodejs.org/api/packages.html#conditional-exports

TL:DR:
- TS resolve `@coveo/cli-commons-dev/what/ever` as `@coveo/cli-commons-dev/lib/what/ever.d.ts`
- JS resolve `@coveo/cli-commons-dev/what/ever` as `@coveo/cli-commons-dev/lib/what/ever.js`


Similar for `@coveo/cli-commons`

File: packages/cli/core/src/commands/config/set.ts [hunk 12]
```diff
@@ -1,7 +1,10 @@
 import {CliUx, Command, Flags} from '@oclif/core';
 import {Config} from '@coveo/cli-commons/config/config';
 import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
-import {IsAuthenticated, Preconditions} from '@coveo/cli-commons/preconditions';
+import {
+  IsAuthenticated,
+  Preconditions,
+} from '@coveo/cli-commons/preconditions/index';
```
Comment (louis-bompart) [inline packages/cli/core/src/commands/config/set.ts:7]: the subpath shenanigans in the packages ain't compatible with folder-as-module.
This is normal, considering that https://nodejs.org/api/modules.html#folders-as-modules is 'legacy'

File: nx.json [hunk 13]
```diff
@@ -10,6 +10,16 @@
...
+  "targetDefaults": {
+    "build": {
+      "dependsOn": ["^build"],
+      "outputs": [
+        "{projectRoot}/lib",
+        "{projectRoot}/dist",
+        "{projectRoot}/tsconfig.build.tsbuildinfo"
+      ]
+    }
+  },
```
Comment (louis-bompart) [inline nx.json:22]: https://nx.dev/reference/project-configuration for global references, line by line comment inc

File: nx.json [hunk 14]
```diff
@@ -10,6 +10,16 @@
...
     "release:phase2": [{"target": "release:phase2", "projects": "dependencies"}]
   },
+  "targetDefaults": {
+    "build": {
+      "dependsOn": ["^build"],
+      "outputs": [
+        "{projectRoot}/lib",
+        "{projectRoot}/dist",
+        "{projectRoot}/tsconfig.build.tsbuildinfo"
+      ]
```
Comment (louis-bompart) [inline nx.json:20]: this is used for the caching of nx. It helps `nx` check/understand if a project is already compiled.

File: packages/cli-e2e/setup/ci.ts [hunk 15]
```diff
@@ -20,6 +20,7 @@ export default async function () {
   if (!(process.env.E2E_USE_NPM_REGISTRY === 'true')) {
     await startVerdaccio();
   }
+  process.env.npm_config_registry = 'http://localhost:4873';
```
Comment (louis-bompart) [inline packages/cli-e2e/setup/ci.ts:23]:
```suggestion
  }
```

**Review by olamothe — APPROVED** [review]
Looks okay, apart from the general fact that there is an awful lot of configuration needed (which I guess is just par for the course for node/TS/JS projects...).

Since you needed to add a bunch of comments on this PR to explain what's what, and why some config is important, what do you think about putting those comments more directly in the projects ?

Some kind of general README for maintainers maybe ? 
That explains how the project is configured, the different tools used, and how it all operates ?

**Pull request merged**